# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 1304 Маркуш Александр. Вариант №41.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

### Аддитивная цепочка
Аддитивная цепочка для натурального числа $n$ – это последовательность натуральных чисел:<br>
$$1=a_0,a_1,a_2,\dots,a_m=n,\quad$$
где каждый элемент получается с помощью суммы каких-либо двух предыдущих:
$$a_i=a_j+a_k, \quad \forall k \leq j < i$$
$l(n) = r$ – наименьшая длина,  для которой существует аддитивная цепочка для $n$.<br>
Звёздной называется цепочка, каждый последующий элемент которой получен следующим образом:<br>
$$a_i=a_{i-1}+a_k\quad \forall k<i$$
Так же существуют следующие оценки длины минимальной аддитивной цепочки числа $n$:
$$l(n) \leq \lambda(n) +\nu(n) - 1$$
$$l(n) \geq \lceil \log_{2}(n) \rceil$$

### Алгоритм Брауэра
&emsp; **Алгоритм Брауера** вычисляет n-ую стпенень за
$\lambda(n)+ \frac{(1 +\: o(1))\:\lambda(n)} {\lambda(\:\lambda(n))}$ операций.
&emsp; Для некоторых $n$, $k$ Брауерские цепочки задаются в виде рекурентной формулы:
$$B_k(n) =\begin{cases}1, 2, 3, ..., 2^k-1\text{, если }n < 2^k \\ B_k(q), 2q, 4q, 8q, ..., 2^kq, n,\text{ если } n \geqslant 2^k\ \text{и } q = \lfloor\frac{n}{2^k}\rfloor \end{cases}$$
В данном алгоритме можно выделить следующие шаги:
<ol>
    <li>Задаётся некий фиксированный пареметр $k$ для $n$. Вычисляем вспомогательные числа:
    $$ d = 2^k,\quad q = [n/d],\quad r_1 = n\mod d,\quad n = q_1d + r_1\quad (0 \leqslant r_1< d)$$</li>
    <li>Певрый пункт проводится, пока не найдётся $q_s < d$. Тогда $q_{s-1} = q_sd + r_s$</li>
    <li>Тогда $n$:
    $$ n = 2^kq_1+r_1 = 2^k(2^kq_2+r_2)+r_1 = \dots = 2^k(2^kq_s + r_s)\dots)+r_2)+r_1$$</li>
</ol>

### Алгоритм дробления вектора индексов
Алгоритм дробления вектора индексов – алгоритм поиска минимальной аддитивной цепочки.
<ol>
    <li>Запускаем внешний цикл по длинам цепочек $ \underline{l}(n) \leq m \leq \overline{l}(n) $. Выбирается натуральное q. Пусть  $ q = m//2$.</li>
    <li>Запускаем внутрненний цикл перебора всех $ \{r_i\}_{i=1}^{q} $. На каждом шаге вычисляем $ a_{min} = a_{q+1} + m - q $ и $ a_{max} = a_{q+1}2^{m-q} $</li>
        <ol>
            <li>Если последний элемент в цепочки равен числу $ n $, заканчиваем алгоритм.</li>
            <li>Если $ n \in [a_{min}, a_{max}] $ , то тогда перебираем часть вектора индексов $ \{r_j\}_{j=q+1}^m $</li>
            <li>Если $ n \notin [a_{min}, a_{max}] $ , то тогда перебираем часть вектора индексов $ \{r_i\}_{i=1}^q $</li>
        </ol>
    <li> Если наборы исчерпаны, увеличиваем длину во внешнем цикле</li>
</ol>

### Гипотеза Шольца-Брауэра
$ l(2^n - 1) \leqslant l(n) + n - 1  $, гипотеза справедлива для $n < 578469$, равенство выполняется для $ n \leqslant 64 $. Для звёздных цепочек гипотеза является доказанной

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики <em>SageMath</em>, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.

## Выполнение работы

### Реализация алгоритма Брауэра
Напишем программу, которая находит аддитивную цепочку для натурального числа $n$ методом Брауэра. Для этого реализована функция
<em>brouwer_algorithm</em>. В ней сначала находится число d, затем найдём вспомогательными числа. Для уменьшения длины цепочки найдём максимальный модуль от деления $r_i = n\mod d$ и построим оставшуюся цепочку вспомогательных чисел до $m//2$ для чётных $m$, и $m//2 + 1$ – для нечётных $m$. Дальше поиск происходит по описанному в теоретических данных алгоритму

In [1]:
def brouwer_algorithm(n, k):
    chain = [1, 2]
    check_points = []
    d = 2**k
    if(n < d):
        for i in range(3, ceil(d/2)+1):
            chain.append(i)
        chain.append(n)
    else:
        while n > d:
            check_points.append(n)
            remainder = n % d
            if remainder not in chain and remainder != 0:
                chain.append(remainder)
            n = n // d
        if n not in chain:
            chain.append(n)
        for i in range(1, ceil(max(chain)/2)+1):
            if i not in chain:
                chain.append(i)
        chain.sort()
        check_points.sort()
        for element in check_points:
            while n*2 < element:
                n *= 2
                if n not in chain:
                    chain.append(n)
            n = element
            if n not in chain:
                    chain.append(n)
        chain.sort()
    return chain


def print_brouwer_algorithm(n, k):
    chain = brouwer_algorithm(n, k)
    print(f"chain = {chain}, len = {len(chain)}")

In [2]:
print_brouwer_algorithm(15, 2)

chain = [1, 2, 3, 6, 12, 15], len = 6


In [3]:
print_brouwer_algorithm(15, 3)

chain = [1, 2, 3, 4, 7, 8, 15], len = 7


In [4]:
print_brouwer_algorithm(15, 4)

chain = [1, 2, 3, 4, 5, 6, 7, 8, 15], len = 9


In [5]:
print_brouwer_algorithm(45, 2)

chain = [1, 2, 3, 4, 8, 11, 22, 44, 45], len = 9


In [6]:
print_brouwer_algorithm(45, 3)

chain = [1, 2, 3, 5, 10, 20, 40, 45], len = 8


In [7]:
print_brouwer_algorithm(45, 4)

chain = [1, 2, 3, 4, 5, 6, 7, 8, 13, 16, 32, 45], len = 12


In [8]:
print_brouwer_algorithm(213, 2)

chain = [1, 2, 3, 6, 12, 13, 26, 52, 53, 106, 212, 213], len = 12


In [9]:
print_brouwer_algorithm(213, 3)

chain = [1, 2, 3, 5, 6, 12, 24, 26, 52, 104, 208, 213], len = 12


In [10]:
print_brouwer_algorithm(213, 4)

chain = [1, 2, 3, 4, 5, 6, 7, 13, 26, 52, 104, 208, 213], len = 13


In [11]:
print_brouwer_algorithm(1024, 2)

chain = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024], len = 11


In [12]:
print_brouwer_algorithm(1024, 3)

chain = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024], len = 11


In [13]:
print_brouwer_algorithm(1024, 4)

chain = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024], len = 11


| n    | k | Цепочка методом Брауэра                          | Длина методом Брауэра | Минимальная цепочка                           | Минимальная длина |
|------|---|--------------------------------------------------|-----------------------|-----------------------------------------------|-------------------|
| 15   | 2 | [1, 2, 3, 6, 12, 15]                             | 6                     | [1, 2, 4, 5, 10, 15]                          | 6                 |
| 15   | 3 | [1, 2, 3, 4, 7, 8, 15]                           | 7                     | [1, 2, 4, 5, 10, 15]                          | 6                 |
| 15   | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 15]                     | 9                     | [1, 2, 4, 5, 10, 15]                          | 6                 |
| 45   | 2 | [1, 2, 3, 4, 8, 11, 22, 44, 45]                  | 9                     | [1, 2, 4, 8, 9, 18, 36, 45]                   | 8                 |
| 45   | 3 | [1, 2, 3, 5, 10, 20, 40, 45]                     | 8                     | [1, 2, 4, 8, 9, 18, 36, 45]                   | 8                 |
| 45   | 4 | [1, 2, 3, 4, 5, 6, 7, 8, 13, 16, 32, 45]         | 12                    | [1, 2, 4, 8, 9, 18, 36, 45]                   | 8                 |
| 213  | 2 | [1, 2, 3, 6, 12, 13, 26, 52, 53, 106, 212, 213]  | 12                    | [1, 2, 4, 8, 16, 32, 33, 49, 82, 164, 213]    | 11                |
| 213  | 3 | [1, 2, 3, 5, 6, 12, 24, 26, 52, 104, 208, 213]   | 12                    | [1, 2, 4, 8, 16, 32, 33, 49, 82, 164, 213]    | 11                |
| 213  | 4 | [1, 2, 3, 4, 5, 6, 7, 13, 26, 52, 104, 208, 213] | 13                    | [1, 2, 4, 8, 16, 32, 33, 49, 82, 164, 213]    | 11                |
| 1024 | 2 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]    | 11                    | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11                |
| 1024 | 3 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]    | 11                    | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11                |
| 1024 | 4 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]    | 11                    | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024] | 11                |

#### Вывод 
Алгоритм Брауэра позволяет быстро найти аддитивную цепочку приближенную к оптимальной. Однако выигрывая в скорости, данный алгоритм не всегда позволяет находить кратчайшую цепочку, и вдобавок, из-за вычисления вспомогательных чисел, которые могут не учавствовать в дальнейшем нахождении аддитивной цепочки, финальная цепочка также получается длинее минимальной.Также результат работы алгоритма зависит от выбранного k, что можно увидеть сравнив результаты в таблице.

### Алгоритм дробления вектора индексов
Была реализована программа находящая кротчайшую аддитивную цепочку методом дробления вектора индексов. Для этого были написаны функции: <em>decrement(vector, q, m)</em>, которая изменяет элементы находящиеся в векторе индексов; <em>get_star_chain(vector)</em>, которая создаёт начальный вектор для каждого внешенего цикла; <em>splitting(n, fixed_m = None)</em> (параметр <em>fixed_m</em> нужен для пункта с проверкой теоремы Шольца-Брауэра), которая реализует сам алгоритм с помощью двух предыдущих функций.

In [14]:
import time


def decrement(vector, q, m):
    for i in range(m, q, -1):        
        if vector[i-1] > 1:
            vector[i-1] -= 1
            return vector
        else: 
            vector[i-1] = i
    return vector
 
    
def get_star_chain(vector):
    chain = [1]
    for i in vector:
        chain.append(chain[-1] + chain[i-1]) 
    return chain


def splitting(n, fixed_m = None):
    for m in range(n.bit_length(), n.bit_length()+bin(n).count('1')):
        if fixed_m:
            m = fixed_m
        vector = [i for i in range(1, m)]
        q = m // 2 - 1
        while True:
            chain = get_star_chain(vector)
            if chain[m-1] == n: 
                return chain
            if n < (chain[q]+m-q) or chain[q]*2**(m-q) < n:
                vector = decrement(vector, 0, q)
                if sum(vector[:q]) == abs(q):
                    break 
            else:
                vector = decrement(vector, q, m-1)
                if sum(vector[q:]) == abs(m-q-1):
                    break
    return chain


def show_splitted_vector(n):
    start_time = time.time()
    chain = splitting(int(n))
    print(f"{time.time() - start_time} seconds")        
    print(f"chain = {chain}, len = {len(chain)}")

In [15]:
print(show_splitted_vector(15))

 0.0007221698760986328 seconds
chain = [1, 2, 4, 5, 10, 15], len = 6
None


In [16]:
print(show_splitted_vector(45))

 0.00841665267944336 seconds
chain = [1, 2, 4, 8, 9, 18, 36, 45], len = 8
None


In [17]:
print(show_splitted_vector(213))

 0.49097633361816406 seconds
chain = [1, 2, 4, 8, 16, 32, 33, 49, 82, 164, 213], len = 11
None


In [18]:
print(show_splitted_vector(1001))

 50.404014587402344 seconds
chain = [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000, 1001], len = 14
None


In [19]:
print(show_splitted_vector(1005))

 50.41733193397522 seconds
chain = [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 201, 402, 804, 1005], len = 14
None


In [20]:
print(show_splitted_vector(1010))

 47.46831202507019 seconds
chain = [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008, 1010], len = 14
None


In [21]:
print(show_splitted_vector(1503))

 306.29884815216064 seconds
chain = [1, 2, 4, 8, 16, 32, 64, 128, 136, 137, 273, 546, 683, 1366, 1503], len = 15
None


In [22]:
print(show_splitted_vector(1024))

 3.886222839355469e-05 seconds
chain = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024], len = 11
None


| n    | Минимальная цепочка                                                | Время работы алгоритма, сек |
|------|--------------------------------------------------------------------|-----------------------------|
| 15   | [1, 2, 4, 5, 10, 15]                                               | 0.0007221698760986328       |
| 45   | [1, 2, 4, 8, 9, 18, 36, 45]                                        | 0.00841665267944336         |
| 213  | [1, 2, 4, 8, 16, 32, 33, 49, 82, 164, 213]                         | 0.49097633361816406         |
| 1001 | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000, 1001]      | 50.404014587402344          |
| 1005 | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 201, 402, 804, 1005]       | 50.41733193397522           |
| 1010 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008, 1010]      | 47.46831202507019           |
| 1503 | [1, 2, 4, 8, 16, 32, 64, 128, 136, 137, 273, 546, 683, 1366, 1503] | 306.29884815216064          |
| 1024 | [1, 2, 4, 8, 16, 32, 33, 49, 82, 164, 213]                         | 3.886222839355469e-05       |

#### Вывод
Алгоритм дробления вектора индексов находит кратчайшую аддитивную цепочку для натурального числа n. Однако за точность алгоритм требует больше времени, нежели рассмотренный ранее метод Брауэра.

### Гипотеза Шольца-Брауэра
Для проверки гипотезы Шольца-Брауэра была написана программа, которая проверяет выполнение гипотезы для $n \leqslant 12$ и выводит результат проверки.

In [27]:
def sholz_brower_theorem(n):
    is_theorem_true = True
    for i in range(1,n+1):
        first_value = len(splitting(int(i))) + i - 1
        second_value = len(splitting(int(2^i - 1), (first_value)))
        if second_value > first_value:
            print("Гипотеза неверна для {second_value}, {first_value}")
            is_theorem_true = False
            break
        else:
            print(f"Теорема выполняется для: {i}")
    if is_theorem_true:
        print("Выполняется для всех заданных n, следовательно гипотеза верна")
        
        
sholz_brower_theorem(12)   

Теорема выполняется для: 1
Теорема выполняется для: 2
Теорема выполняется для: 3
Теорема выполняется для: 4
Теорема выполняется для: 5
Теорема выполняется для: 6
Теорема выполняется для: 7
Теорема выполняется для: 8
Теорема выполняется для: 9
Теорема выполняется для: 10
Теорема выполняется для: 11
Теорема выполняется для: 12
Выполняется для всех заданных n, следовательно гипотеза верна


#### Вывод
Была подтверждена гипотеза Шольца-Брауэра для значений n, не превышающих 12. 

## Выводы
В ходе работы приобретены знания о теории аддитивных цепочек, а также наработаны практические навыки составления и применения алгоритмов для определения минимальных аддитивных цепочек для заданных чисел. Были освоены навыки работы с SageMath для реализации алгоритмов. Также были реализованы и проверены алгоритмы Брауэра и дробления вектора индексов, а результаты сравнены между собой. Была проверена гипотеза Шольца-Брауэра для $n \leqslant 12$.